File: b49/sdk/src/crdt.ts
import { v4 as uuidv4 } from 'uuid';
import type { Dot, TaskItem, Task, TaskList, CreateTaskOptions, TaskItemOptions } from './types';

export function createDot(replicaId: string, counter: number): Dot {
  return {
    replica_id: replicaId,
    counter
  };
}

export function dotKey(dot: Dot): string {
  return `${dot.replica_id}:${dot.counter}`;
}

export function dotEquals(a: Dot, b: Dot): boolean {
  return a.replica_id === b.replica_id && a.counter === b.counter;
}

export function dotsEqual(a: Dot[], b: Dot[]): boolean {
  if (a.length !== b.length) return false;
  const setA = new Set(a.map(dotKey));
  return b.every(dot => setA.has(dotKey(dot)));
}

export function taskItemKey(item: TaskItem): string {
  return item.id;
}

export function taskKey(task: Task): string {
  return task.id;
}

export function serializeTaskItem(item: TaskItem): string {
  return JSON.stringify({
    id: item.id,
    content: item.content,
    completed: item.completed,
    created_at: item.created_at,
    updated_at: item.updated_at
  });
}

export function deserializeTaskItem(key: string, data: string): TaskItem {
  return JSON.parse(data) as TaskItem;
}

export function serializeTask(task: Task): string {
  return JSON.stringify({
    id: task.id,
    title: task.title,
    description: task.description,
    items: task.items,
    created_at: task.created_at,
    updated_at: task.updated_at
  });
}

export function deserializeTask(key: string, data: string): Task {
  return JSON.parse(data) as Task;
}

export function createORSet<T>(
  getKey: (item: T) => string,
  serialize: (item: T) => string,
  deserialize: (key: string, data: string) => T
) {
  return {
    elements: {} as Record<string, Dot[]>,
    tombstones: {} as Record<string, Dot[]>,
    getKey,
    _serialize: serialize,
    _deserialize: deserialize,

    add(value: T, dot: Dot): void {
      const key = getKey(value);
      const serialized = serialize(value);
      if (!this.elements[serialized]) {
        this.elements[serialized] = [];
      }
      if (!this.elements[serialized].some(d => dotEquals(d, dot))) {
        this.elements[serialized].push(dot);
      }
    },

    remove(value: T, dot: Dot): void {
      const serialized = serialize(value);
      if (this.elements[serialized]) {
        const dots = [...this.elements[serialized]];
        delete this.elements[serialized];

        if (!this.tombstones[serialized]) {
          this.tombstones[serialized] = [];
        }
        for (const d of dots) {
          if (!this.tombstones[serialized].some(td => dotEquals(td, d))) {
            this.tombstones[serialized].push(d);
          }
        }
        if (!this.tombstones[serialized].some(td => dotEquals(td, dot))) {
          this.tombstones[serialized].push(dot);
        }
      }
    },

    contains(value: T): boolean {
      const serialized = serialize(value);
      return !!this.elements[serialized];
    },

    iter(): T[] {
      return Object.entries(this.elements).map(([serialized, _]) => 
        deserialize(getKey(JSON.parse(serialized) as T), serialized)
      );
    },

    len(): number {
      return Object.keys(this.elements).length;
    },

    isEmpty(): boolean {
      return this.len() === 0;
    },

    merge(other: any): void {
      for (const [serialized, dots] of Object.entries(other.elements || {})) {
        if (!this.elements[serialized]) {
          this.elements[serialized] = [];
        }
        for (const dot of dots as Dot[]) {
          if (!this.elements[serialized].some(d => dotEquals(d, dot))) {
            this.elements[serialized].push(dot);
          }
        }
      }

      for (const [serialized, dots] of Object.entries(other.tombstones || {})) {
        if (!this.tombstones[serialized]) {
          this.tombstones[serialized] = [];
        }
        for (const dot of dots as Dot[]) {
          if (!this.tombstones[serialized].some(d => dotEquals(d, dot))) {
            this.tombstones[serialized].push(dot);
          }
        }
      }

      const elementKeys = Object.keys(this.elements);
      for (const serialized of elementKeys) {
        const tombstoneDots = this.tombstones[serialized];
        const elementDots = this.elements[serialized];
        
        if (tombstoneDots && elementDots) {
          const allCovered = elementDots.every(ed => 
            tombstoneDots.some(td => dotEquals(ed, td))
          );
          if (allCovered) {
            delete this.elements[serialized];
          }
        }
      }
    }
  };
}

export function createTaskItemORSet() {
  return createORSet<TaskItem>(taskItemKey, serializeTaskItem, deserializeTaskItem);
}

export function createTaskORSet() {
  return createORSet<Task>(taskKey, serializeTask, deserializeTask);
}

export function createTask(options: CreateTaskOptions, replicaId: string): { task: Task; dot: Dot } {
  const now = new Date().toISOString();
  const task: Task = {
    id: uuidv4(),
    title: options.title,
    description: options.description || null,
    items: createTaskItemORSet() as any,
    created_at: now,
    updated_at: now
  };
  const dot = createDot(replicaId, 0);
  return { task, dot };
}

export function createTaskItem(options: TaskItemOptions): TaskItem {
  const now = new Date().toISOString();
  return {
    id: uuidv4(),
    content: options.content,
    completed: options.completed ?? false,
    created_at: now,
    updated_at: now
  };
}

export function updateTaskItem(item: TaskItem, updates: { content?: string; completed?: boolean }): TaskItem {
  return {
    ...item,
    ...(updates.content !== undefined ? { content: updates.content } : {}),
    ...(updates.completed !== undefined ? { completed: updates.completed } : {}),
    updated_at: new Date().toISOString()
  };
}

export function updateTask(task: Task, updates: { title?: string; description?: string | null }): Task {
  return {
    ...task,
    ...(updates.title !== undefined ? { title: updates.title } : {}),
    ...(updates.description !== undefined ? { description: updates.description } : {}),
    updated_at: new Date().toISOString()
  };
}

export function createTaskList(): TaskList {
  return {
    tasks: createTaskORSet() as any,
    last_sync_at: null
  };
}

export function mergeTaskLists(local: TaskList, remote: TaskList): TaskList {
  const result: TaskList = {
    tasks: createTaskORSet() as any,
    last_sync_at: new Date().toISOString()
  };
  
  result.tasks.merge(local.tasks);
  result.tasks.merge(remote.tasks);
  
  return result;
}

File: b49/backend/src/crdt.rs
use std::collections::{HashMap, HashSet};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct Dot {
    pub replica_id: String,
    pub counter: u64,
}

impl Dot {
    pub fn new(replica_id: &str, counter: u64) -> Self {
        Self {
            replica_id: replica_id.to_string(),
            counter,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ORSet<T>
where
    T: std::hash::Hash + Eq + Clone + Serialize + for<'de> Deserialize<'de>,
{
    pub elements: HashMap<T, HashSet<Dot>>,
    pub tombstones: HashMap<T, HashSet<Dot>>,
}

impl<T> ORSet<T>
where
    T: std::hash::Hash + Eq + Clone + Serialize + for<'de> Deserialize<'de>,
{
    pub fn new() -> Self {
        Self {
            elements: HashMap::new(),
            tombstones: HashMap::new(),
        }
    }

    pub fn add(&mut self, value: T, dot: Dot) {
        self.elements
            .entry(value)
            .or_insert_with(HashSet::new)
            .insert(dot);
    }

    pub fn remove(&mut self, value: T, dot: Dot) {
        if let Some(dots) = self.elements.remove(&value) {
            let tombstones = self
                .tombstones
                .entry(value)
                .or_insert_with(HashSet::new);
            tombstones.extend(dots);
            tombstones.insert(dot);
        }
    }

    pub fn contains(&self, value: &T) -> bool {
        self.elements.contains_key(value)
    }

    pub fn iter(&self) -> impl Iterator<Item = &T> {
        self.elements.keys()
    }

    pub fn len(&self) -> usize {
        self.elements.len()
    }

    pub fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }

    pub fn merge(&mut self, other: &Self) {
        for (elem, dots) in &other.elements {
            let entry = self
                .elements
                .entry(elem.clone())
                .or_insert_with(HashSet::new);
            entry.extend(dots.iter().cloned());
        }

        for (elem, dots) in &other.tombstones {
            let tombstone_entry = self
                .tombstones
                .entry(elem.clone())
                .or_insert_with(HashSet::new);
            tombstone_entry.extend(dots.iter().cloned());
        }

        let elements_keys: Vec<T> = self.elements.keys().cloned().collect();
        for elem in elements_keys {
            if let Some(tombstone_dots) = self.tombstones.get(&elem) {
                if let Some(element_dots) = self.elements.get(&elem) {
                    if tombstone_dots.is_superset(element_dots) {
                        self.elements.remove(&elem);
                    }
                }
            }
        }
    }
}

impl<T> Default for ORSet<T>
where
    T: std::hash::Hash + Eq + Clone + Serialize + for<'de> Deserialize<'de>,
{
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskItem {
    pub id: String,
    pub content: String,
    pub completed: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: String,
    pub title: String,
    pub description: Option<String>,
    pub items: ORSet<TaskItem>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Task {
    pub fn new(title: String, description: Option<String>, replica_id: &str) -> (Self, Dot) {
        let now = Utc::now();
        let task = Self {
            id: Uuid::new_v4().to_string(),
            title,
            description,
            items: ORSet::new(),
            created_at: now,
            updated_at: now,
        };
        let dot = Dot::new(replica_id, 0);
        (task, dot)
    }

    pub fn add_item(&mut self, item: TaskItem, dot: Dot) {
        self.items.add(item, dot);
        self.updated_at = Utc::now();
    }

    pub fn remove_item(&mut self, item: &TaskItem, dot: Dot) {
        self.items.remove(item.clone(), dot);
        self.updated_at = Utc::now();
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskList {
    pub tasks: ORSet<Task>,
    pub last_sync_at: Option<DateTime<Utc>>,
}

impl TaskList {
    pub fn new() -> Self {
        Self {
            tasks: ORSet::new(),
            last_sync_at: None,
        }
    }

    pub fn add_task(&mut self, task: Task, dot: Dot) {
        self.tasks.add(task, dot);
    }

    pub fn remove_task(&mut self, task: &Task, dot: Dot) {
        self.tasks.remove(task.clone(), dot);
    }

    pub fn merge(&mut self, other: &Self) {
        self.tasks.merge(&other.tasks);
        self.last_sync_at = Some(Utc::now());
    }

    pub fn get_task(&self, task_id: &str) -> Option<&Task> {
        self.tasks.iter().find(|t| t.id == task_id)
    }
}

impl Default for TaskList {
    fn default() -> Self {
        Self::new()
    }
}
